### PR TITLE
refactor(developer): hashmap-based wordlist compilation

### PR DESCRIPTION
**Pull request by jahorton**
A follow-up from a TODO in #3338.

### CODE REVIEW

File: developer/js/source/lexical-model-compiler/build-trie.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { readFileSync } from "fs";
  * A word list is an array of pairs: the concrete word form itself, followed by
```
Comment (darcywong00) [inline developer/js/source/lexical-model-compiler/build-trie.ts:4]: Is word list now an Object?

File: developer/js/source/lexical-model-compiler/build-trie.ts [hunk 2]
```diff
@@ -243,8 +233,9 @@ namespace Trie {
      * @param words a list of word and count pairs.
```
Comment (darcywong00) [inline developer/js/source/lexical-model-compiler/build-trie.ts:233]: Confirming `count` and `weight` are the same thing

File: developer/js/tests/test-parse-wordlist.ts [hunk 3]
```diff
@@ -4,39 +4,39 @@ import 'mocha';
...
-  ['SEN', 4816],
-  ['Ȼ', 3479],
-  ['SW̱', 2621],
-  ['NIȽ', 2314],
-  ['U¸', 2298],
-  ['I¸', 1988],
-  ['ȻSE', 1925],
-  ['I', 1884]
-];
+const SENCOTEN_WORDLIST = {
```
Comment (darcywong00) [inline developer/js/tests/test-parse-wordlist.ts:7]: Is this now a `WordList` type?

Comment (darcywong00) [reply developer/js/tests/test-parse-wordlist.ts:7]: I guess a complication is we couldn't define a const that had identical wordforms (e.g.)
```
'I': 2, // from wordlist 1
'I', 5, // from wordlist 2
...
```

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 1 above)
Comment (jahorton) [reply developer/js/source/lexical-model-compiler/build-trie.ts:4]: The terminology gets tricky here.  The new `WordList` definition technically is rooted in a JavaScript object, yes.  However, the technical term for this pattern is [_associative array_]().

That said, even if we kept the "array" term, it's no longer an array of pairs.  It's more like a Java `HashMap`.

File: developer/js/tests/test-parse-wordlist.ts (hunk 3 above)
Comment (jahorton) [reply developer/js/tests/test-parse-wordlist.ts:7]: Yep.  This matches the new format specified for `WordList` [here](https://github.com/keymanapp/keyman/pull/3340/files#diff-b0a389458ccb6df98f56754ae35edd35R7).

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 1 above)
Comment (mcdurdin) [reply developer/js/source/lexical-model-compiler/build-trie.ts:4]: > Java

Well, actually... :rofl:

File: developer/js/source/lexical-model-compiler/build-trie.ts (hunk 2 above)
Comment (jahorton) [reply developer/js/source/lexical-model-compiler/build-trie.ts:233]: I believe they're correlated, yes.  The word count is used to determine the weight.

File: developer/js/tests/test-parse-wordlist.ts (hunk 3 above)
Comment (mcdurdin) [reply developer/js/tests/test-parse-wordlist.ts:7]: > I guess a complication is we couldn't define a const that had identical wordforms (e.g.)

The final wordlist should never have two entries with the same wordform: that's the whole point of the parent PR 😄

**Comment by mcdurdin** [comment]
I like this. I didn't want to make this change in the initial PR because I was trying to keep it very targeted. But this is also pretty neat and tidy 😀.

Have you done any performance testing? I'd be keen to see a before/after on nrc.mtnt.en for instance.

**Comment by mcdurdin** [comment]
@eddieantonio Would you mind checking this one as well when you look at the parent #3338?

**Comment by jahorton** [comment]
OK, makeshift performance test results are in:

**Base branch**
before - stress-test

**This branch**
after - stress-test

I used [gnomon](https://www.npmjs.com/package/gnomon) to time it.  It seems that it put the test's runtime as a prefix on the previous line.  (As if it's timing how long each line in the console stood as 'most recent'.)

So... 136s -> 9s.  Seems like an improvement to me.

**Comment by mcdurdin** [comment]
> So... 136s -> 9s. Seems like an improvement to me.

So it's about 15 times faster. Okay, that'll do, tiger, that'll do. 🐅

File: developer/js/tests/test-parse-wordlist.ts [hunk 4]
```diff
@@ -69,17 +69,16 @@ describe('parseWordList', function () {
       ' hello ', //4, expect to trim whitespace
       'hello']; //5
 
-    const expected = [
-      [ 'hello', 10 /* 1+4+5 trimmed and identical */ ],
-      [ 'hell'+String.fromCharCode(0x00F3), 5 /* 2+3 normalised to NFC */ ]
-    ];
+    const expected: WordList = {};
+    expected['hello'] = 10; /* 1+4+5 trimmed and identical */
+    expected['hell'+String.fromCharCode(0x00F3)] = 5 /* 2+3 normalised to NFC */
```
Comment (eddieantonio) [inline developer/js/tests/test-parse-wordlist.ts:74]: Why not use a literal here?

```suggestion
    const expected: WordList = {
        'hello': 10, /* 1+4+5 trimmed and identical */
        'hell\u00f3': 5, /* 2+3 normalised to NFC */
    };
```

Comment (mcdurdin) [reply developer/js/tests/test-parse-wordlist.ts:74]: Great idea -- wanted to make it clear that it was an NFC form ... this is more concise.

**Pull request merged**